### PR TITLE
ROS#60 - HAL design

### PR DESCRIPTION
**Pull request by Glamhoth**

### CODE REVIEW

File: aerugo-hal/src/system_hal.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+use core::ops::{Add, Sub};
+
+/// System HAL trait.
+pub trait SystemHal {
+    /// Type for an instant in time.
+    type Instant: Ord
+        + Copy
+        + Add<Self::Duration, Output = Self::Instant>
+        + Sub<Self::Duration, Output = Self::Instant>
+        + Sub<Self::Instant, Output = Self::Duration>;
```
Comment (SteelPh0enix) [inline aerugo-hal/src/system_hal.rs:17]: Why no `Add<Self::Instant, Output = Self::Duration>`? Is it automatically generated somehow?

File: aerugo-hal/src/system_hal.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    /// Gets current system time timestamp.
+    fn get_system_time(&self) -> Self::Instant;
+
+    /// Feeds the system watchdog.
+    fn feed_watchdog(&mut self);
+
+    /// Enters critical section
+    fn enter_critical()
+    where
+        Self: Sized;
```
Comment (SteelPh0enix) [inline aerugo-hal/src/system_hal.rs:38]: kill if not needed

File: aerugo-hal/src/system_hal.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    /// Exits critical section
+    fn exit_critical()
+    where
+        Self: Sized;
```
Comment (SteelPh0enix) [inline aerugo-hal/src/system_hal.rs:43]: kill if not needed

File: aerugo-hal/src/system_hal.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    /// * `F` - Closure type.
+    /// * `R` - Closure return type.
+    ///
+    /// * `f` - Closure to execute.
+    ///
+    /// Returns closure result.
+    fn execute_critical<F, R>(f: F) -> R
+    where
+        F: FnOnce(&CriticalSection) -> R,
+        Self: Sized;
```
Comment (SteelPh0enix) [inline aerugo-hal/src/system_hal.rs:56]: kill if not needed

File: arch/cortex-m/samv71-hal/src/hal.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    pub const fn new(peripherals: Peripherals) -> Self {
+        Hal {
+            _peripherals: peripherals,
+        }
+    }
+}
+
+impl SystemHal for Hal {
+    type Instant = crate::time::TimerInstantU64<1_000_000>;
+    type Duration = crate::time::TimerDurationU64<1_000_000>;
```
Comment (SteelPh0enix) [inline arch/cortex-m/samv71-hal/src/hal.rs:25]: move magic number to named constant (if it's related to system clock or configuration, it should later be made tailorable via config or something like that)

File: arch/x86/x86-hal/src/hal.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    pub const fn new(peripherals: Peripherals) -> Self {
+        Hal {
+            _peripherals: peripherals,
+        }
+    }
+}
+
+impl SystemHal for Hal {
+    type Instant = crate::time::TimerInstantU64<1_000_000>;
+    type Duration = crate::time::TimerDurationU64<1_000_000>;
```
Comment (SteelPh0enix) [inline arch/x86/x86-hal/src/hal.rs:32]: same as in V71 HAL

File: aerugo-hal/src/system_hal.rs (hunk 1 above)
Comment (Glamhoth) [reply aerugo-hal/src/system_hal.rs:17]: Adding `Duration` to `Instant` you still get an `Instant`.
image

File: aerugo-hal/src/system_hal.rs (hunk 2 above)
Comment (Glamhoth) [reply aerugo-hal/src/system_hal.rs:38]: done

File: aerugo-hal/src/system_hal.rs (hunk 3 above)
Comment (Glamhoth) [reply aerugo-hal/src/system_hal.rs:43]: done

File: aerugo-hal/src/system_hal.rs (hunk 4 above)
Comment (Glamhoth) [reply aerugo-hal/src/system_hal.rs:56]: done

File: arch/cortex-m/samv71-hal/src/hal.rs (hunk 5 above)
Comment (Glamhoth) [reply arch/cortex-m/samv71-hal/src/hal.rs:25]: I'm not 100% sure how does configuring that would look like, so you can treat this as a stub (as this is part of the implementation, not the API). We will do some form of configuration/having this as a common element across the system when we will tackle the TC driver.

File: arch/x86/x86-hal/src/hal.rs (hunk 6 above)
Comment (Glamhoth) [reply arch/x86/x86-hal/src/hal.rs:32]: Same answer as above.

File: arch/cortex-m/samv71-hal/src/hal.rs (hunk 5 above)
Comment (SteelPh0enix) [reply arch/cortex-m/samv71-hal/src/hal.rs:25]: is it possible to at least move it to a separate constant? It seems like both of these values are supposed to be equal, so it'd make sense to take them from the same source

**Review by SteelPh0enix — DISMISSED** [review]

File: arch/cortex-m/samv71-hal/src/hal.rs [hunk 7]
```diff
@@ -0,0 +1,56 @@
+//! System HAL implementation for x86 target.
+
+use aerugo_hal::system_hal::{SystemHal, SystemHardwareConfig};
+use bare_metal::CriticalSection;
+
+use crate::peripherals::Peripherals;
+
+/// HAL implementation for x86.
```
Comment (Lurkerpas) [inline arch/cortex-m/samv71-hal/src/hal.rs:8]: in samv71-hal?

File: arch/x86/x86-hal/src/hal.rs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+                .elapsed()
+                .expect("{}")
+                .as_nanos()
+                .try_into()
+                .unwrap(),
+        )
+    }
+
+    fn feed_watchdog(&mut self) {
+        todo!()
```
Comment (Lurkerpas) [inline arch/x86/x86-hal/src/hal.rs:53]: Are we planning to implement watchdog on x86? We don't have HW, but we COULD have SW watchtog in another thread....

Comment (Lurkerpas) [reply arch/x86/x86-hal/src/hal.rs:53]: But again, given that x86 on Linux is not real-time, does it make sense? We did have a watchdog for some Android SW, with period counted in minutes.

File: arch/cortex-m/samv71-hal/src/hal.rs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            _peripherals: peripherals,
+        }
+    }
+}
+
+impl SystemHal for Hal {
+    type Instant = crate::time::TimerInstantU64<{ Hal::TIMER_FREQ }>;
+    type Duration = crate::time::TimerDurationU64<{ Hal::TIMER_FREQ }>;
+
+    fn configure_hardware(&mut self, _config: SystemHardwareConfig) {
```
Comment (Lurkerpas) [inline arch/cortex-m/samv71-hal/src/hal.rs:30]: I assume  master clock is part of the platform-specific SystemHardwareConfig?

File: arch/x86/x86-hal/src/hal.rs (hunk 8 above)
Comment (Glamhoth) [reply arch/x86/x86-hal/src/hal.rs:53]: I think that there is no point really, as we are only doing the x86 target for development purposes. I just added `todo` everywhere as a stub, but I can replace this one as we won't implement that at all.

File: arch/cortex-m/samv71-hal/src/hal.rs (hunk 7 above)
Comment (Glamhoth) [reply arch/cortex-m/samv71-hal/src/hal.rs:8]: fixed

File: arch/cortex-m/samv71-hal/src/hal.rs (hunk 9 above)
Comment (Glamhoth) [reply arch/cortex-m/samv71-hal/src/hal.rs:30]: That is the plan, yes.

**Review by SteelPh0enix — APPROVED** [review]

**Pull request merged**